Modules/_collectionsmodule.c

```diff
@@ -1904,15 +2006,25 @@ dequeiter_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
 static PyObject *
 dequeiter_len(dequeiterobject *it, PyObject *Py_UNUSED(ignored))
 {
-    return PyLong_FromSsize_t(it->counter);
+    Py_ssize_t len = _Py_atomic_load_ssize(&it->counter);
```

Comment:
For atomic operation, can you adopt same approach that I suggested on dict and list to reduce side-effect for default build?
https://github.com/python/cpython/pull/114843/files

cc @erlend-aasland

Reply:
The benchmark show no significant slowdown. Do you think it is worth the added complexity, @corona10?

Reply:
> The benchmark show no significant slowdown. Do you think it is worth the added complexity, @corona10?

IMO, In some cases, it will be okay to use atomic operation, and in some cases, it will not be okay; we can not run every benchmark for every change during the first milestone of PEP 703.
So I would like to suggest avoiding such micro benchmark as possible, we can measure them once the PEP 703 is completed.
Once we switch the default build to use the disabled GIL version, we can naturally switch them to use atomic operation, while the GIL version does not need to worry about the changes from PEP 703.

And from my observation, it only needs seamless load operation and store operation and I think that it is not that complex, it's less complex than we have to write fragmented logic if we think that it will increase the overhead to the default build.

Reply:
Thanks for explaining your rationale and motivation :)

Reply:
In this change, we only need something like beacuse `_Py_atomic_load_ssize` only used.

```
#ifdef Py_GIL_DISABLED
    #define LOAD_SSIZE_ATOMIC_AS_POSSIBLE(value) _Py_atomic_load_ssize(&value)
#else
    #define LOAD_SSIZE_ATOMIC_AS_POSSIBLE(value) value
#endif
```

Reply:
> If I understand you correctly, you're suggesting to add an internal macro for this somewhere in Include/?

Not sure at this moment, but I think that it will be better.

Reply:
>  all atomic loads in this PR?

All

> because there is a lot of atomic operations in the CPython source code already,

That's true, but it does not mean that we are okay to change into atomic operations aggressively where the non-atomic operation was used since if there was GIL, it is not needed.
If we only care about free-threaded build, it's fine, but the situation is not.

I know that this is too conservative an approach, but we do not need to take any risk since it is technically possible to maintain both non-atomic operation and atomic operation in the same code line that does not break any readability.

Reply:
>  and most of them are not special cased like this.

IIUC, in the GIL build, most of them are related to memory ordering issues where the atomic operations are used.

Reply:
So, for me, this discussion is about strategy issues rather than technical issues.
(I also bet that there will be low possibility of performance effect to default build through this change.)

1. How do we determine that the switching to atomic operation is safe for the default build?
2. Do we have to measure every operation when the change is needed?
3. How about just skipping the performance measurement while developing PEP 703 and maintaining both operations? And once the first milestone is completed, then we measure the overall performance.

Reply:
Nobody asked me, but my 2 cents :) 

1. We should use a macro (I'd prefer a static inline function) like the one @corona10 suggested when we need to replace a non-atomic load/store with an atomic one for free-threaded builds. In many cases we'll want to do this to avoid the overhead of introducing a critical section just to perform a load/store. Since the critical section would be a no-op in the default build, we should try to make whatever we use to replace it also a no-op in default builds.
2. Whatever we use (macro or function) should follow the naming convention used for existing atomic helpers. That is, it should use the most conservative memory ordering (`seq_cst`) unless either the function name indicates otherwise (e.g. `XXX_load_relaxed`) or the memory order is supplied as a parameter.
3. We should add this recommendation to documentation somewhere (dev guide?) along with the naming convention for functions that expect to be called with per-object locks held.

Reply:
The macros would not be my first choice, but I'm okay with them. The relaxed atomic loads and stores compile to plain memory loads and stores -- we don't need to benchmark every use to know that.

Reply:
> Nobody asked me, but my 2 cents :)

Thanks for chiming in, both of you! I was just trying to understand Donghee's rationale for the requested change; now I do :)

I trust you three will find a good solution here :)

Reply:
> Whatever we use (macro or function) should follow the naming convention used for existing atomic helpers. That is, it should use the most conservative memory ordering (seq_cst) unless either the function name indicates otherwise (e.g. XXX_load_relaxed) or the memory order is supplied as a parameter.

Yeah, we need some convention. Do you want to work for utility APIs?
It will be useful for other runtime migration tasks too. (My PR is also need and I will use it)
For preference between macro and static inline function, I prefer macro side because this is just one-line simple code.
but I would like to delegate the decision to @mpage and @colesbury to use the macro strategy or inline functions strategy.

Thank you for listening to my grumpy conservative stands :)

cc @erlend-aasland

Reply:
>Do you want to work for utility APIs?

Sure I can take a stab at it.